base/sort.jl

```diff
@@ -1111,6 +1119,197 @@ function _sort!(v::AbstractVector, a::ScratchQuickSort, o::Ordering, kw;
+  * *in-place* in memory if the `next` algorithm is in-place.
+  * *estimate-and-filter*: strategy
+  * *linear runtime* if `length(target)` is constant and `next` is reasonable
+  * *n + k log k* worst case runtime if `next` has that runtime.
+  * *pathological inputs* can significantly increase constant factors.
+"""
+struct BracketedSort{T, F} <: Algorithm
+    target::T
+    get_next::F
+end
+
+# TODO: this composition between BracketedSort and ScratchQuickSort does not bring me joy
+BracketedSort(k::Integer) = BracketedSort(k, k -> InitialOptimizations(PartialQuickSort(k)))
+BracketedSort(k::OrdinalRange) = BracketedSort(k, k -> InitialOptimizations(ScratchQuickSort(k)))
+
+function bracket_kernel!(v::AbstractVector, lo, hi, lo_signpost, hi_signpost, o)
+    i = 0
+    count_below = 0
+    checkbounds(v, lo:hi)
+    for j in lo:hi
+        x = @inbounds v[j]
+        a = lo_signpost !== nothing && lt(o, x, lo_signpost)
+        b = hi_signpost === nothing || !lt(o, hi_signpost, x)
+        count_below += a
+        # if a != b # This branch is almost never taken, so making it branchless is bad.
+        #     @inbounds v[i], v[j] = v[j], v[i]
+        #     i += 1
+        # end
+        c = a != b # JK, this is faster.
+        k = i * c + j
+        # Invariant: @assert firstindex(v) ≤ lo ≤ i + j ≤ k ≤ j ≤ hi ≤ lastindex(v)
+        @inbounds v[j], v[k] = v[k], v[j]
+        i += c - 1
+    end
+    count_below, i+hi
+end
+
+function move!(v, target, source)
+    # This function never dominates runtime—only add `@inbounds` if you can demonstrate a
+    # performance improvement. And if you do, also double check behavior when `target`
+    # is out of bounds.
+    @assert length(target) == length(source)
+    if length(target) == 1 || isdisjoint(target, source)
+        for (i, j) in zip(target, source)
+            v[i], v[j] = v[j], v[i]
+        end
+    else
+        @assert minimum(source) <= minimum(target)
+        reverse!(v, minimum(source), maximum(target))
+        reverse!(v, minimum(target), maximum(target))
+    end
+end
+
+function _sort!(v::AbstractVector, a::BracketedSort, o::Ordering, kw)
+    @getkw lo hi scratch
+    # TODO for further optimization: reuse scratch between trials better, from signpost
+    # selection to recursive calls, and from the fallback (but be aware of type stability,
+    # especially when sorting IEEE floats.
+
+    # We don't need to bounds check target because that is done higher up in the stack
+    # However, we cannot assume the target is inbounds.
+    lo < hi || return scratch
+    ln = hi - lo + 1
+
+    # This is simply a precomputed short-circuit to avoid doing scalar math for small inputs.
+    # It does not change dispatch at all.
+    ln < 260 && return _sort!(v, a.get_next(a.target), o, kw)
+
+    target = a.target
+    k2 = round(Int, ln^(2/3))
+    k2ln = k2/ln
+    offset = .7k2^0.575 # TODO for further optimization: tune this
```

Comment:
When something is well enough optimized but could be a little faster, I'm leaving a note so I or others can revisit these things to squeeze out the last drops of performance. A `# TODO for further optimization` need not ever be addressed for an algorithm to run real fast.

Reply:
IMO, this should be at least significantly simplified if it's not tuned. 2 floating point pows will be fairly expensive so it IMO requires being pretty good to be worth using such heavy weight computation.

Reply:
Somewhat complext calculations are necessary for the asymptotic runtime of the algorithm. A sample size of `k2` gives an expected accuracy of within `length/sqrt(k2)`, so to get the minimum expected value for `sample size + elements wrongfully included` one needs to make `k2` about `length ^ (2/3)`. `offset` determines exactly how much to compensate to make sure we get the target in the selected range almost all the time. I bumped up the exponent from the theoretical `1/2` to `.575` so that the odds of success are very high for large inputs when the lower order terms become negligible and I really want to avoid having to make an additional pass through the input. We could use `isqrt` or some other faster way of computing these results, but It only runs on inputs of length > 260, so the constant factor isn't too bad (max about 2% of total runtime, typically much less) so I figured I'd leave the computations in the form that feels most natural from an algorithmic design perspective.

Does that seem like an adequate justification or do you still think this should be simpler or more computationally efficient?

Reply:
That makes sense. The 2 things I would recommend then are probably to change the logic to
```
tmp = cbrt(ln)
k2 = round(Int, tmp^2)
offset = .35*tmp*top_set_bit(k2)
```
`cbrt(x)` is much faster than `^`, and this changes the `offset` to a formula based on `sqrt(k2)*log(k2)` which is fast enough to ensure the chance decays to 0 while being faster to compute.

Reply:
Sounds good. `k2` stands for `k^2`, so I'll call `tmp` `k`, and .15 is a better constant than `.35`. For lengths between `260` and `typemax(Int)`, this changes the offset by a factor of 0.9391292915807987 to 1.5971905796294013 and on small inputs of length 260 to 1000, where it matters more, it changes the offset by a factor of 0.9391292915807987 to 1.1023369386283894.

And the heuristic is 5-6x faster to compute now